File: reg_tools/csv2reg.py
'''
Script used to automatically register UCC students.
---
Created Date: Sunday November 20th 2022
Author: Jefferson Ding
-----
Last Modified: Sunday November 20th 2022 11:08:20 pm
Modified By: Jefferson Ding at <jefferson.ding@ucc.on.ca>
-----
Copyright (c) 2023 Jefferson Ding
'''
import firebase_admin
from firebase_admin import credentials
from firebase_admin import db
from firebase_admin import auth

import uuid
import csv
import json

from firebase import Firebase
with open("api.json", "r") as f:
    config = json.load(f)
    firebase = Firebase(config)

# initialize firebase
cred = credentials.Certificate("./key.json")
firebase_admin.initialize_app(cred, {"databaseURL": "https://worldaffairscon-8fdc5-default-rtdb.firebaseio.com"})
# initialize realtime database
db = firebase_admin.db
# initialize firebase authentication
auth = firebase_admin.auth

# registration constants
access_code = "THOZYirJHLYj9UXKnypCs9vbeTw1"


# get students from csv
def get_students(csv_path):
    students = []
    with open(csv_path, 'r') as file:
        reader = csv.DictReader(file)
        for row in reader:
            students.append(row)
    return students


# register student to firebase auth
def firebase_auth_register(email):
    temp_password = str(uuid.uuid4().hex)
    try:
        user = auth.create_user(
            email=email,
            email_verified=False,
            password=temp_password,
            disabled=False
        )
        print(f"Successfully created new user: {user.uid} with email: {email} and temporary password: {temp_password}")
    except Exception as e:
        print(f"Error creating new user: {e}")
        return None
    return user.uid

def add_user(email,name,grade,access_code,ucc_advisor,id):
    uid = firebase_auth_register(email)
    if uid is not None:
        # set student reference
        ref = "teachers/" + access_code + "/students/" + uid
        db.reference(ref).set({
            "email": email,
            "name": name,
            "grade": grade,
            "p1": "",
            "p2": "",
            "p3": "",
            "ucc_advisor": ucc_advisor,
            "note": "",
            "uccid": id
        })

        # set teacher link
        ref = "students/" + uid
        db.reference(ref).set({
            "teacherID": access_code
        })
        reset_password(email)
    else:
        print("Failed to register user due to firebase auth error")

def reset_password(email):
    firebase.auth().send_password_reset_email(email)
    print("Password reset email sent to " + email)

if __name__=="__main__":
    students = get_students("./students2.csv")
    for student in students:
        email = student["email"]
        name = student["pre_first_name"] + " " + student["last_name"]
        grade = student["grade"].replace("Year ", "")
        ucc_advisor = student["advisor"]
        studentnum = student["id"]
        add_user(email,name,grade,access_code,ucc_advisor,studentnum)

    

File: 2023/csv2reg.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"strings"

	firebase "firebase.google.com/go"
	"firebase.google.com/go/auth"
	"github.com/google/uuid"
	"golang.org/x/net/context"
	"google.golang.org/api/option"
)

var accessCode = "THOZYirJHLYj9UXKnypCs9vbeTw1"

func importUCC() {
	// Initialize Firebase app
	opt := option.WithCredentialsFile("key.json")
	config := &firebase.Config{
		DatabaseURL: "https://worldaffairscon-8fdc5-default-rtdb.firebaseio.com",
	}
	ctx := context.Background()
	app, err := firebase.NewApp(ctx, config, opt)

	log.Fatalln("error initializing app:", err)

	client, err := app.Auth(ctx)

	log.Fatalln("error initializing Auth:", err)

	if err != nil {
		log.Fatalln("error initializing app:", err)
	}

	// Get students from csv
	students, err := getStudents("./test.csv")
	if err != nil {
		log.Fatalln("Error getting students from csv:", err)
	}

	// Register students to Firebase auth
	for _, student := range students {
		email := student["email"]
		name := student["pre_first_name"] + " " + student["last_name"]
		grade := strings.Replace(student["grade"], "Year ", "", 1)
		uccAdvisor := student["advisor"]
		studentNum := student["id"]
		temp := uuid.NewString()
		fmt.Println(email, name, grade, uccAdvisor, studentNum, temp)
		params := (&auth.UserToCreate{}).
			Email(email).
			Password(temp).
			Disabled(false)
		u, err := client.CreateUser(ctx, params)
		if err != nil {
			log.Fatalf("error creating user: %v\n", err)
		}
		log.Printf("Successfully created user: %v with email %v, and temporary password %v", u, email, temp)
	}
}

func getStudents(csvPath string) ([]map[string]string, error) {
	file, err := os.Open(csvPath)
	if err != nil {
		return nil, fmt.Errorf("error opening csv file: %v", err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	records, err := reader.ReadAll()
	if err != nil {
		return nil, fmt.Errorf("error reading csv file: %v", err)
	}

	students := make([]map[string]string, len(records)-1)
	for i, record := range records[1:] {
		student := make(map[string]string)
		for j, field := range record {
			student[records[0][j]] = field
		}
		students[i] = student
	}

	return students, nil
}

func sendPasswordResetEmail(email string) error {

	return nil
}

func Use(vals ...interface{}) {
	for _, val := range vals {
		_ = val
	}
}
